lib/AST/Printer/StmtPrinter.cpp

```diff
@@ -3202,40 +3202,76 @@ PrintedTokenRange PrintedTokenRange::Create(clang::ASTContext &context,
   auto tokens = std::make_shared<PrintedTokenRangeImpl>(context);
 
   if (stmt) {
+    PrintingPolicyAdaptor ppa;
+    tokens->ppa = &ppa;
     StmtPrinter printer(out, nullptr, *tokens, policy);
     printer.Visit(stmt);
+    tokens->ppa = nullptr;
   }
 
+  tokens->tokens.emplace_back(
+      0u, 0u, kInvalidTokenContextIndex, 0u, 0u, clang::tok::eof);
+
   auto num_tokens = tokens->tokens.size();
-  if (!num_tokens) {
-    return PrintedTokenRange(std::move(tokens));
-  } else {
-    auto first = &(tokens->tokens[0]);
-    auto after_last = &(first[num_tokens]);
-    return PrintedTokenRange(std::move(tokens), first, after_last);
-  }
+  auto first = &(tokens->tokens[0]);
+  auto after_last = &(first[num_tokens - 1u]);
+
+  return PrintedTokenRange(std::move(tokens), first, after_last);
 }
 
 PrintedTokenRange PrintedTokenRange::Create(const std::shared_ptr<ASTImpl> &ast,
-                                            clang::Stmt *stmt) {
+                                            clang::Stmt *stmt,
+                                            const PrintingPolicy &high_pp) {
   std::string data;
   raw_string_ostream out(data, 0);
   auto &context = ast->tu->getASTContext();
   auto tokens = std::make_shared<PrintedTokenRangeImpl>(context);
 
+  // Top-level context should be the AST.
+  tokens->ast = ast;
+  tokens->contexts.emplace_back(*ast);
+
   if (stmt) {
+    PrintingPolicyAdaptor ppa(ast, high_pp);
+    tokens->ppa = &ppa;
+
     clang::PrintingPolicy pp = *(ast->printing_policy);
+    pp.IncludeTagDefinition = high_pp.ShouldPrintTagBodies();
+
     StmtPrinter printer(out, nullptr, *tokens, pp);
     printer.Visit(stmt);
+    tokens->ppa = nullptr;
   }
 
+  // Fixup to share the AST as the root context.
+  auto max_i = tokens->contexts.size();
+  for (auto i = 1u; i < max_i; ++i) {
+    TokenContextImpl &context = tokens->contexts[i];
+    if (context.parent_index == kInvalidTokenContextIndex) {
+      context.parent_index = kASTTokenContextIndex;
+    }
+  }
+
+  tokens->tokens.emplace_back(
+      0u, 0u, kInvalidTokenContextIndex, 0u, 0u, clang::tok::eof);
+
   auto num_tokens = tokens->tokens.size();
-  if (!num_tokens) {
-    return PrintedTokenRange(std::move(tokens));
-  } else {
-    auto first = &(tokens->tokens[0]);
-    auto after_last = &(first[num_tokens]);
-    return PrintedTokenRange(std::move(tokens), first, after_last);
+  auto first = &(tokens->tokens[0]);
```

Comment:
Ah, this has more code because this variant links back to a PASTA AST, and thus there is a defined "AST" token context, whereas the other doesn't. Should this distinction go in a comment?